### DISCUSSION TITLE
Sending midi events to different types of endpoints

### DISCUSSION THREAD

**Discussion by fracturedsoftware** [orchetect/swift-midi#180] [Q&A] [answered]
Is MIDIKit able to create a connection that can send to both an internal input as well as an output that external devices can connect to?

Using MIDIKit I can send an event to a single managedOutputConnection that has multiple managedInputs and inputsUnowned attached to it, but I can’t figure out a way to also send to one or more outputs that may or may not have another device listening to it. The only way seems to be to send the same events twice:

A bit like:

    midiManager.addOutputConnection(
        toInputs: [],
        toOutputs:[]
        tag: "Test2_Output"
    )

For example - is there any way to combine the event sending at the end of the following?

```
/// Create an output connection
try midiManager?.addOutputConnection( toInputs: [], tag: "Test2_Output" )
outputConnection = midiManager?.managedOutputConnections["Test2_Output"]

/// create an input and add it to the outputConnection
try? midiManager.addInput( name: "Keyboard", tag: "Keyboard_Tag", uniqueID: .adHoc, receiver: .events (keyboard.eventHandler))
let input = midiManager.managedInputs["Keyboard_Tag"]
outputConnection?.add(inputs: [.uniqueID((input!.uniqueID!))])

/// Add any unowned inputs
DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
    self.midiManager.endpoints.inputsUnowned.forEach({
        self.outputConnection?.add(inputs:  [.uniqueID(($0.uniqueID))])
    })
}

/// Send an event to all inputs
DispatchQueue.main.asyncAfter(deadline: .now() + 0.2) {
    let event =  MIDIEvent.noteOn( 60, velocity: .unitInterval(0.5), channel: 0, group: 0 )
    try? self.outputConnection?.send(event: event)
}

/// Create a virtual output and send an event to it
try midiManager?.addOutput( name: "MIDIKit Event Logger Out", tag: "EventLoggerOutput", uniqueID: .adHoc )
let output = self.midiManager?.managedOutputs["EventLoggerOutput"]

DispatchQueue.main.asyncAfter(deadline: .now() + 0.2) {
    try? output?.send(event: .start())
}

/// Send events to both <- Combine these into one send?
DispatchQueue.main.asyncAfter(deadline: .now() + 0.2) {
    let event =  MIDIEvent.noteOn( 60, velocity: .unitInterval(0.5), channel: 0, group: 0 )
    try? self.outputConnection?.send(event: event)
    try? output?.send(event: event)
}
```

Thanks!

**Comment by orchetect** [comment, marked as answer]
No, an output connection and an output endpoint are separate concepts and are addressed individually.

I'd probably recommend just making a single method you can call that will send the same event to both.

I'm curious what the use case is here (the overall goal you're trying to achieve). Maybe there's a simpler way to implement.

Also, if you're going to reuse any of your virtual endpoints, I would not use the `.adHoc` ID. It's better to use `.userDefaultsManaged` and assign it a key so its ID can be restored.
